src/Microsoft.OData.Core/UriParser/SemanticAst/ODataPathExtensions.cs

```diff
@@ -137,6 +137,22 @@ public static ODataPath TrimEndingTypeSegment(this ODataPath path)
             return handler.FirstPart;
         }
 
+        /// <summary>
+        /// Creates a <see cref="ODataPath"/> that is <paramref name="path"/> with the type segments and key segments removed from the end
+        /// </summary>
+        /// <param name="path">The <see cref="ODataPath"/> to trim the ending of</param>
+        /// <returns>A <see cref="ODataPath"/> without type-cast and key segments at the end</returns>
+        /// <exception cref="System.ArgumentNullException">Thrown if <paramref name="path"/> is <see langword="null"/></exception>
+        public static ODataPath TrimEndingTypeAndKeySegments(this ODataPath path)

```

Comment:
Do we have unit tests for this?

Reply:
I think we had good coverage with the JSON writer, but you're totally right, it's better to have direct coverage of the extension, thanks!